more_itertools/more.py

```diff
@@ -3121,37 +3123,48 @@ def replace(iterable, pred, substitutes, count=None, window_size=1):
         >>> list(replace(iterable, pred, substitutes, window_size=window_size))
         [3, 4, 5, 3, 4, 5]
 
+    Use *default_arg* to specify the default value passed as arguments to
+    *pred* when at the end of the iterable.
+    This option is only used when *window_size* is greater than 1.
+
+        >>> list(replace(
+        ...     [1, 0, 0, 5, 0, 1, 1],
+        ...     # 3 items passed to pred with default of 0
+        ...     lambda *args: args == (1, 0, 0),
+        ...     [3, 4], # Splice in these items
+        ...     window_size=3,
+        ...     default_arg=0,
+        ... ))
+        [3, 4, 5, 0, 1, 3, 4]
+
     """
     if window_size < 1:
         raise ValueError('window_size must be at least 1')
 
     # Save the substitutes iterable, since it's used more than once
     substitutes = tuple(substitutes)
 
-    # Add padding such that the number of windows matches the length of the
-    # iterable
-    it = chain(iterable, [_marker] * (window_size - 1))
-    windows = windowed(it, window_size)
+    # Create the iterable with padding so we don't run out of elements
+    # to pass to the predicate
+    iterable = chain(iter(iterable), (default_arg,) * (window_size - 1))
+
+    # Create an almost full window
+    window = deque(islice(iterable, window_size - 1), maxlen=window_size)
+
+    # An iterable that will advance the window
+    filler = map(window.append, iterable)
 
     n = 0
-    for w in windows:
-        # If the current window matches our predicate (and we haven't hit
-        # our maximum number of replacements), splice in the substitutes
-        # and then consume the following windows that overlap with this one.
-        # For example, if the iterable is (0, 1, 2, 3, 4...)
-        # and the window size is 2, we have (0, 1), (1, 2), (2, 3)...
-        # If the predicate matches on (0, 1), we need to zap (0, 1) and (1, 2)
-        if pred(*w):
-            if (count is None) or (n < count):
-                n += 1
-                yield from substitutes
-                consume(windows, window_size - 1)
-                continue
-
-        # If there was no match (or we've reached the replacement limit),
-        # yield the first item from the window.
-        if w and (w[0] is not _marker):
-            yield w[0]
+    for _ in filler:
+        # Produce substitues and advance the window if
+        # the predicate is true and we haven't reached our count
+        if pred(*window) and (count is None or n < count):
```

Comment:
```suggestion
        if (count is None or n < count) and pred(*window):
```
Swapping the evaluation order would speed up cases where `count` replacements have been happened.